Use iv.py to display images in terminal

MatanZ opened the pull request:
This replaces kitty discovery and terminal image display with a module I wrote to support graphics terminal.

This module supports three graphics protocol (sixel, iterm2, kitty), and recognizes support by relevant escape sequences, rather than by checking terminal name.

I tested this commit in konsole (22.04), xterm, wezterm and kitty. It should work in any terminal that supports any of those protocols.

MatanZ committed:
Use iv.py to display images in terminal

dotcs commented:
Thanks a ton for those changes, @MatanZ! It will take me a few days to look through and understand the proposed changes, so please stay tuned.

dotcs commented:
Thanks again for this PR! While reviewing it, I found a couple of issues we should have a look into. I've merged changes from `develop` into this branch that introduces a GitHub workflow to automatically lint and run a static code analysis on the branches. This should also happen automatically on the next commit in this PR. Meanwhile, can you please run those two commands manually and fix the issues they report? Code formatting can be done automatically with `black` (see [docs](https://pypi.org/project/black/)) while `pyright` found a couple of typing issues that need to be resolved.

```
$ poetry run python -m black --check --diff .
$ poetry run python -m pyright
```

Please also specify in detail if you have used other code for this PR. I found [this code](https://github.com/sandkoan/dotfiles/blob/45eac26b222e5c89153b0b54563f4508acd4377a/scripts/png.py#L4-L13) to be very similar to the `serialize_gr_command` function in this PR. Eventually please double check that any submitted code in this PR is compliant with the current license (MIT).

In the meantime I try to better understand the code and the mechanics behind it. ;)

MatanZ commented:
> Please also specify in detail if you have used other code for this PR. I found [this code](https://github.com/sandkoan/dotfiles/blob/45eac26b222e5c89153b0b54563f4508acd4377a/scripts/png.py#L4-L13) to be very similar to the serialize_gr_command function in this PR. Eventually please double check that any submitted code in this PR is compliant with the current license (MIT).

This is from [here](https://sw.kovidgoyal.net/kitty/graphics-protocol/#a-minimal-example). I never considered the license, but considering it is a small example in what the author hopes will become a standard, I expect it to be freelt reusable without limitation

MatanZ committed:
- Replace KBhit class with specific implementation of read input with timeout
- Replace serialize_gr_command with more straightforward writing.
- Lint using `black`
- Reorder methods a bit

MatanZ commented:
I believe this addresses all your comments. I removed the example code from kitty with my own, so no question of copyright. Similarly for the KBhit class.

I also corrected the issues pointed to by pyright.

MatanZ committed:
Need to `black` xkcd.py as well

dotcs committed:
Add typing information

dotcs committed:
Remove unused imports

dotcs commented:
Thanks for updating the PR. In the meantime I think I got a better understanding of those changes and of the ANSI escape codes that you have used. I'll add a couple of commits to this PR in order to improve typing support, and I'll add some docstrings. Maybe we can also add some tests for the critical parts of the code. Please don't force-push any longer in this PR, so that we both can commit without issues. :+1:

dotcs committed:
Add dependency: pillow

dotcs committed:
Improve typings

dotcs committed:
Check if 'convert' is installed before using it

dotcs committed:
Add docstrings

dotcs committed:
Add option to explicily set width of image

MatanZ committed:
Cosmetic changes

- Reformat help strings
- Replace 'kitty' with 'terminal graphics' (or similar) where appropriate.

MatanZ commented:
Why are some of the typer help strings multilined with `"""`, and another with `("", "")`?

MatanZ committed:
Change scaling logic:

Default to fit to screen. Override with `--width`.
Disable up scaling with `--no-terminal-scale-up`.

dotcs commented:
> Why are some of the typer help strings multilined with `"""`, and another with `("", "")`?

Thanks for the hint. That's not intentional. We should stick with a single way to multiline strings. The latter one is probably more convenient because there are no issues with whitespace at the beginning of lines, so I would like to pick this one. WDYT?

MatanZ committed:
Need to run `black`

dotcs reviewed (commented):
Thanks for fixing the wrong behavior and [your detailed explanation](https://github.com/dotcs/xkcd-cli/pull/3#commitcomment-75014161). :+1: 

With this implementation I have an issue when using a terminal that falls back to the `convert` command. `poetry run xkcd show --comic-id 218 --width 100` triggers the cmd `['convert', '/tmp/tmpck1g5hmr/218.png', '-geometry', '100x-1', 'sixel:-']` under the hood which will not work, because the geometry `100x-1` is not defined. I guess instead `100x` should be used.

Should we cover this case specifically or is this a general problem with the `-1` value also in other branches?

dotcs commented on xkcd_cli/iv.py:
```diff
@@ -0,0 +1,356 @@
+                    print()
+
+    # Get various sizes of screen
+    def pixel_size(self) -> Tuple[int, int]:
```
Should we rename this method to `terminal_window_size` or something like this? I think `pixel_size` is a bit misleading as it might also refer to a synthetic pixel size within the terminal window or something like this.

MatanZ replied:
It is not exactly the window size. It is the size of the area that may be covered by graphics (which is usually what may be covered by text). The terminal might have other widgets in the window, might split the window to multiple terminals, etc.

And size is not descriptive enough. the size may be measured in pixels or characters (and in the future, may be mm or similar). Maybe terminal_pixel_size and rename terminal_size to terminal_cell_size.

dotcs committed:
Add docstrings

dotcs commented on xkcd_cli/iv.py:
```diff
@@ -0,0 +1,407 @@
+        """
+        if self.ex_kitty is None:
+            # A small jpeg from https://stackoverflow.com/questions/2253404
+            self.ex_kitty = "OK" in self.terminal_request(
+                "\x1b_Gi=31,s=1,v=1,a=q,t=d,f=100;"
+                + "/9j/2wBDAAEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQH/wgALCAABAAEBAREA/8QAFAABAAAAAAAAAAAAAAAAAAAAA//aAAgBAQAAAAE//9k="
+                + "\x1b\\",
+                "\\",
+            )
```
@MatanZ For me it's still not clear how you came up with this idea. I understand that you're rendering a PNG (`f=100`) image with 1x1 dimensions `s=1,v=1` in query action mode (`a=q`). But I don't understand how you came up with this specific payload. Can you please add a comment to explain how you did this so that in case anything breaks later have a better understanding of how this works? Thanks a ton in advance.

dotcs commented on xkcd_cli/iv.py:
```diff
@@ -0,0 +1,407 @@
+        """
+        if self.ex_kitty is None:
+            # A small jpeg from https://stackoverflow.com/questions/2253404
+            self.ex_kitty = "OK" in self.terminal_request(
+                "\x1b_Gi=31,s=1,v=1,a=q,t=d,f=100;"
+                + "/9j/2wBDAAEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQH/wgALCAABAAEBAREA/8QAFAABAAAAAAAAAAAAAAAAAAAAA//aAAgBAQAAAAE//9k="
+                + "\x1b\\",
+                "\\",
+            )
```
@MatanZ For me it's still not clear how you came up with this idea. I understand that you're rendering a PNG (`f=100`) image with 1x1 dimensions `s=1,v=1` in query action mode (`a=q`). But I don't understand how you came up with this specific payload. Can you please add a comment to explain how you did this so that in case anything breaks later have a better understanding of how this works? Thanks a ton in advance.

MatanZ replied:
This is not a PNG, but a JPG. Kitty supports pngs only and such is defined in the documentation. But konsole and wezterm also accept jpegs. This code tests if jpeg is supported.

For xkcd_cli this is unnecessary, as the source is a png anyway. But in general, if the source image is jpeg, sending a jpeg saves a lot of traffic (I usually use terminal graphics over ssh on a slow link).

I need to make clear: `iv.py` is a python module which I intend to distribute independently to allow python program to display graphics in terminal. So it does include a few features that are not useful to `xkcd-cli`. Though the might be useful in the future if you decide to support other comics which are distributed as jpegs.

dotcs replied:
I see. Now it makes a lot more sense. Thanks for explaining. And btw, I like the idea of making a python package out of that! :+1:

dotcs commented on xkcd_cli/iv.py:
```diff
@@ -0,0 +1,407 @@
+        return ret
+
+    # Methods to send show image in various protocols.
+    def kitty_remove_placement(self, p: int = -1) -> None:
+        sys.stdout.buffer.write(f"\033_Ga=d,i=-1,d=z,z={p},q=2\033\\".encode("ascii"))
+        sys.stdout.flush()
```
@MatanZ This method is not in use any longer. Can we safely remove it?

MatanZ replied:
It was never used here, and there is no problem with removing it.. See my explanation above.

It may be used in the future if you want to make the program interactive (display comic and then the user presses 'n' to display the next comic). This function allows you to remove the image without scrolling it out (and maybe to display it later without sending it to the terminal again).

dotcs replied:
Great, again thanks for explaining. I'll keep it here in case we can make use of it later on.

dotcs committed:
Add comments to better explain kitty control data keys

dotcs committed:
Merge branch 'develop' into iv

dotcs committed:
Update changelog

dotcs committed:
Update changelog

dotcs committed:
Update readme

MatanZ committed:
Correct geometry in call to ImagemMagick's `convert`.

MatanZ commented:
This corrects sixel behaviour with `convert`. 

There is still an issue with libsixel, as this library has no option for keeping aspect ratio, so if given both width and height (as we do now), it fits exactly, changing the aspect of the image.

dotcs committed:
Rename functions; improve docstrings

dotcs committed:
Add/improve docstrings

dotcs committed:
Add dev dependency: pytest

dotcs committed:
Refactor fn 'terminal_request' to make it better testable

dotcs committed:
Add some tests

dotcs committed:
Add tests to gh workflow

dotcs committed:
Add more tests

dotcs committed:
Add dev dependency: pytest-cov

dotcs committed:
Isolate extracting img data and metadata with PIL in separate method

dotcs committed:
Add tests for show_image method; refactor method to make it better
testable

dotcs committed:
Add tests for sixel fallback mechanism

dotcs committed:
Add test for libpixel path

dotcs committed:
Add testcase: no libsixel, no convert

dotcs committed:
Add tests for jpg/png image and metadata read fn

MatanZ committed:
Improve width and height argument handling in libsixel case.

dotcs committed:
Add tests for 'kitty_show_file' method

dotcs committed:
Use parameterized tests for 'test_scale_fit' method

dotcs committed:
Add tests for iterm specific fns

dotcs reviewed (approved):
I've added a couple of tests to ensure that most critical parts are working as expected and we have some kind of documentation for the behavior. Would be great if you can look through the tests and let me know if you spot any issues, @MatanZ. Otherwise I'd consider this PR to be ready to merge. Again thanks a ton for your hard work! :partying_face:

MatanZ commented:
I am from the _if it compiles, ship it_ camp, so I know very little about testing infrastructure and such things. The functions do seem like they test what is expected, but I do not see anything that uses the functions in iv_test.py. I expected there to be the python equivalent of `make test` which runs the tests and reports success.

dotcs commented:
> I am from the if it compiles, ship it camp, so I know very little about testing infrastructure and such things.

I'm more from the camp that tries to sleep well by ensuring that the code in production runs. Not for all things, but at least for the more complex parts. I still have to add tests to the original code base though. 😂

Sorry for not being more specific on how to execute the tests. For now please run `poetry run python -m pytest --cov="xkcd_cli/" --cov-report term --cov-report html` to run all tests and create a html report of the code coverage. You might need to update packages with `poetry install` first, since I added some packages while working on this PR.

dotcs committed:
Add test for identification of supported protocols

dotcs committed:
Merge branch 'develop' into iv
